File: amb_cli/integrations/stitch/stitch_client.mjs
#!/usr/bin/env node
/**
 * 🎨 AMB_V2 - Runner Node.js do Google Stitch SDK (@google/stitch-sdk)
 * Localização: amb_v2/integrations/stitch/stitch_client.mjs
 * Responsabilidade Única: Estabelecer conexão com o StitchToolClient e despachar chamadas oficiais.
 */

import { StitchToolClient, stitch } from '@google/stitch-sdk';
import fs from 'fs';
import path from 'path';

function getEnvValue(key) {
  if (process.env[key]) return process.env[key].trim().replace(/^['"]|['"]$/g, '');

  let curr = process.cwd();
  for (let i = 0; i < 6; i++) {
    const envFile = path.join(curr, '.env');
    if (fs.existsSync(envFile)) {
      try {
        const txt = fs.readFileSync(envFile, 'utf-8');
        for (const line of txt.split('\n')) {
          const t = line.trim();
          if (t.startsWith(`${key}=`)) {
            return t.split('=')[1].trim().replace(/^['"]|['"]$/g, '');
          }
        }
      } catch {}
    }
    const parent = path.dirname(curr);
    if (parent === curr) break;
    curr = parent;
  }
  return null;
}

function getApiKey() {
  return getEnvValue('STITCH_API_KEY') || getEnvValue('GEMINI_API_KEY');
}

function getProjectId() {
  return getEnvValue('STITCH_PROJECT_ID');
}

async function extractScreenDetails(client, projectId, result) {
  let screenId = null;
  let screenshotUrl = null;
  let htmlCode = null;
  let title = "";
  let description = "";
  const messages = [];

  if (result && typeof result === 'object') {
    if (result.screenId) screenId = result.screenId;
    if (result.name && typeof result.name === 'string') screenId = result.name.split('/').pop();
    if (result.screenshot?.downloadUrl) screenshotUrl = result.screenshot.downloadUrl;
    if (result.screenshotUrl) screenshotUrl = result.screenshotUrl;
    if (result.htmlCode?.downloadUrl) htmlCode = result.htmlCode.downloadUrl;
    if (typeof result.htmlCode === 'string') htmlCode = result.htmlCode;
    if (result.title) title = result.title;
    if (result.description) description = result.description;
    if (result.message) messages.push(result.message);

    const comps = result.outputComponents || result.output_components || [];
    for (const comp of comps) {
      if (comp.text) messages.push(comp.text.trim());
      if (comp.design?.screens) {
        for (const s of comp.design.screens) {
          if (s.name) screenId = s.name.split('/').pop();
          if (s.screenshot?.downloadUrl) screenshotUrl = s.screenshot.downloadUrl;
          if (s.htmlCode?.downloadUrl) htmlCode = s.htmlCode.downloadUrl;
          if (s.title) title = s.title;
          if (s.description) description = s.description;
        }
      }
    }
  }

  if (htmlCode && typeof htmlCode === 'string' && (htmlCode.startsWith('http://') || htmlCode.startsWith('https://'))) {
    try {
      const resp = await fetch(htmlCode);
      if (resp.ok) {
        htmlCode = await resp.text();
      }
    } catch {}
  }

  return {
    success: true,
    screenId,
    screenshotUrl,
    htmlCode,
    title,
    description,
    messages
  };
}

async function main() {
  const args = process.argv.slice(2);
  if (args.length < 2) {
    console.error(JSON.stringify({
      error: "Argumentos insuficientes.",
      usage: "node stitch_client.mjs <command> '<json_payload>'"
    }));
    process.exit(1);
  }

  const [command, rawPayload] = args;
  let payload = {};
  try {
    payload = JSON.parse(rawPayload);
  } catch (e) {
    console.error(JSON.stringify({ error: `Payload JSON inválido: ${e.message}` }));
    process.exit(1);
  }

  const apiKey = getApiKey();
  if (!apiKey) {
    console.error(JSON.stringify({
      error: "STITCH_API_KEY ausente.",
      hint: "Configure STITCH_API_KEY no arquivo .env"
    }));
    process.exit(1);
  }

  if (!payload.projectId) {
    const envPid = getProjectId();
    if (envPid) payload.projectId = envPid;
  }

  if (apiKey) {
    process.env.STITCH_API_KEY = apiKey;
  }

  try {
    const client = new StitchToolClient({ apiKey });
    let response;

    switch (command) {
      case "generate_screen":
      case "generate_screen_from_text": {
        const res = await client.callTool("generate_screen_from_text", payload);
        response = await extractScreenDetails(client, payload.projectId, res);
        break;
      }
      case "generate_variants": {
        const varPayload = { ...payload };
        if (!varPayload.selectedScreenIds && varPayload.screenId) {
          varPayload.selectedScreenIds = [varPayload.screenId];
          delete varPayload.screenId;
        }
        if (!varPayload.variantOptions) {
          varPayload.variantOptions = {
            variantCount: varPayload.variantCount || 3,
            creativeRange: varPayload.creativeRange || "EXPLORE"
          };
        }
        const res = await client.callTool("generate_variants", varPayload);
        response = await extractScreenDetails(client, payload.projectId, res);
        break;
      }
      case "edit_screens":
      case "edit_screen": {
        const editPayload = { ...payload };
        if (!editPayload.selectedScreenIds && editPayload.screenId) {
          editPayload.selectedScreenIds = [editPayload.screenId];
          delete editPayload.screenId;
        }
        const res = await client.callTool("edit_screens", editPayload);
        response = await extractScreenDetails(client, payload.projectId, res);
        break;
      }
      case "get_screen": {
        const screenName = payload.name || `projects/${payload.projectId}/screens/${payload.screenId}`;
        const res = await client.callTool("get_screen", {
          name: screenName,
          projectId: payload.projectId,
          screenId: payload.screenId
        });
        response = await extractScreenDetails(client, payload.projectId, res);
        try {
          const projResp = await client.callTool("get_project", { name: `projects/${payload.projectId}` });
          const proj = projResp?.project || projResp;
          if (proj?.designTheme?.designMd) {
            response.designMd = proj.designTheme.designMd;
          }
        } catch {}
        break;
      }
      case "get_project": {
        const projName = payload.name || (payload.projectId ? `projects/${payload.projectId}` : (payload.id ? `projects/${payload.id}` : ''));
        response = await client.callTool("get_project", { name: projName });
        break;
      }
      case "create_project": {
        response = await client.callTool("create_project", payload || {});
        break;
      }
      case "list_projects": {
        response = await client.callTool("list_projects", payload || {});
        break;
      }
      case "list_screens": {
        const rawScreensResp = await client.callTool("list_screens", { projectId: payload.projectId });
        const screensList = rawScreensResp?.screens || [];
        
        try {
          const projName = `projects/${payload.projectId}`;
          const projResp = await client.callTool("get_project", { name: projName });
          const instances = projResp?.project?.screenInstances || projResp?.screenInstances || [];
          
          const screenMap = new Map();
          screensList.forEach(s => {
            const sid = s.id || (s.name ? s.name.split('/').pop() : '');
            if (sid) screenMap.set(sid, s);
          });

          for (const inst of instances) {
            const sid = inst.id || (inst.sourceScreen ? inst.sourceScreen.split('/').pop() : '');
            if (sid && !screenMap.has(sid)) {
              screenMap.set(sid, {
                id: sid,
                name: inst.sourceScreen || `projects/${payload.projectId}/screens/${sid}`,
                title: inst.label || inst.title || `Tela ${sid.substring(0, 8)}`,
                description: inst.description || (inst.type === 'DESIGN_SYSTEM_INSTANCE' ? 'Instância de Design System' : ''),
                width: inst.width,
                height: inst.height,
                hidden: inst.hidden
              });
            } else if (sid && screenMap.has(sid) && inst.label) {
              screenMap.get(sid).title = inst.label;
            }
          }
          response = { screens: Array.from(screenMap.values()) };
        } catch (e) {
          response = rawScreensResp;
        }
        break;
      }
      case "create_design_system": {
        response = await client.callTool("create_design_system", payload);
        break;
      }
      case "update_design_system": {
        response = await client.callTool("update_design_system", payload);
        break;
      }
      case "list_design_systems": {
        response = await client.callTool("list_design_systems", payload || {});
        break;
      }
      case "apply_design_system": {
        response = await client.callTool("apply_design_system", payload);
        break;
      }
      case "download_assets": {
        response = await client.callTool("download_assets", payload);
        break;
      }
      case "upload":
      case "upload_asset": {
        const proj = stitch.project(payload.projectId);
        const screens = await proj.upload(payload.filePath, payload.opts || {});
        response = {
          success: true,
          screens: (screens || []).map(s => ({
            id: s.id,
            screenId: s.screenId,
            projectId: s.projectId
          }))
        };
        break;
      }
      default: {
        response = await client.callTool(command, payload);
        break;
      }
    }

    console.log(JSON.stringify(response, null, 2));
    try { await client.close(); } catch {}
  } catch (err) {
    console.error(JSON.stringify({
      error: err.message || String(err),
      details: err.stack
    }));
    process.exit(1);
  }
}

main();

File: amb_cli/integrations/stitch/stitch_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
🎨 AMB_V2 - Google Stitch SDK Client & Tooling (SRP)
Localização: amb_v2/integrations/stitch/stitch_client.py
Responsabilidade Única: Interface Python unificada para geração de telas, refinamentos,
variantes visuais, listagem de telas e sincronização de Design Tokens via Stitch Runner Node.js.
"""

import os
import sys
import json
import shutil
import subprocess
from typing import Dict, Any, List, Optional

from core import (
    Colors,
    log,
    log_error,
    require_env,
    get_env,
    ApiExecutionError,
    ConfigurationError,
)
from workspace import get_device_type
from amb_cli.integrations.stitch.stitch_core.asset_manager import (
    download_assets_core,
    upload_asset_core,
    save_screen_html_core,
)
from amb_cli.integrations.stitch.stitch_core.project_manager import (
    get_project_core,
    create_project_core,
    list_projects_core,
)
from amb_cli.integrations.stitch.stitch_core.design_sync import (
    create_design_system_core,
    update_design_system_core,
    list_design_systems_core,
    apply_design_system_core,
    sync_design_system_core
)


class StitchClient:
    """Client oficial para o Google Stitch SDK (@google/stitch-sdk)."""

    def __init__(self, project_id: Optional[str] = None):
        self.api_key = require_env("STITCH_API_KEY")
        self.project_id = project_id or get_env("STITCH_PROJECT_ID")
        self.runner_path = self._resolve_runner()

    def _resolve_runner(self) -> str:
        candidates = [
            os.path.abspath(os.path.join(os.path.dirname(__file__), "stitch_client.mjs")),
            os.path.abspath(os.path.join(os.path.dirname(__file__), "client", "stitch_client.mjs")),
        ]
        runner = next((p for p in candidates if os.path.exists(p)), None)
        if not runner:
            raise ApiExecutionError(f"Runner Stitch (stitch_client.mjs) não encontrado em {candidates[0]}")
        return runner

    def _run_node_command(self, action: str, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Executa comando no runner Node.js do Stitch SDK com diagnóstico preventivo de runtime."""
        node_bin = shutil.which("node")
        if not node_bin:
            raise ConfigurationError(
                "Runtime Node.js não foi encontrado no PATH do sistema.",
                hint="Instale o Node.js (v18+) para utilizar as funcionalidades do Stitch SDK."
            )

        proc = subprocess.run(
            [node_bin, self.runner_path, action, json.dumps(payload)],
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace"
        )
        if proc.returncode != 0:
            err_msg = proc.stderr.strip() or proc.stdout.strip()
            if "MODULE_NOT_FOUND" in err_msg or "Cannot find module" in err_msg:
                raise ConfigurationError(
                    "Dependência @google/stitch-sdk não encontrada no módulo do Stitch.",
                    hint="Execute 'npm install' no diretório 'integrations/stitch/' para instalar o SDK oficial."
                )
            try:
                err_json = json.loads(err_msg)
                err_msg = err_json.get("error", err_msg)
            except Exception:
                pass
            raise ApiExecutionError(f"Falha na execução do Stitch SDK ({action}): {err_msg}")

        try:
            return json.loads(proc.stdout.strip())
        except json.JSONDecodeError:
            return {"output": proc.stdout.strip()}

    save_screen_html = staticmethod(save_screen_html_core)

    def generate_screen(
        self,
        prompt: str,
        device_type: Optional[str] = None,
        model_id: Optional[str] = None,
        design_system: Optional[str] = None,
        output_file: Optional[str] = None
    ) -> Dict[str, Any]:
        """Gera uma nova tela visual a partir de uma descrição textual respeitando a preferência de dispositivo."""
        project_id = self.project_id or require_env("STITCH_PROJECT_ID")
        resolved_device = (device_type or get_device_type())
        payload = {
            "projectId": project_id,
            "prompt": prompt
        }
        if model_id:
            payload["modelId"] = model_id
        if resolved_device:
            payload["deviceType"] = resolved_device
        if design_system:
            payload["designSystem"] = design_system

        dev_label = f" ({resolved_device})" if resolved_device else ""
        log("STITCH", f"Disparando geração de nova tela visual{dev_label}...", Colors.CYAN)
        res = self._run_node_command("generate_screen_from_text", payload)
        if output_file and res.get("htmlCode"):
            self.save_screen_html(res, output_file)
        return res

    def edit_screen(
        self,
        screen_id: str,
        prompt: str,
        device_type: Optional[str] = None,
        model_id: Optional[str] = None,
        output_file: Optional[str] = None
    ) -> Dict[str, Any]:
        """Refina e edita uma tela existente com novas instruções visuais."""
        project_id = self.project_id or require_env("STITCH_PROJECT_ID")
        resolved_device = (device_type or get_device_type())
        payload = {
            "projectId": project_id,
            "selectedScreenIds": [screen_id],
            "prompt": prompt
        }
        if resolved_device:
            payload["deviceType"] = resolved_device
        if model_id:
            payload["modelId"] = model_id

        log("STITCH", f"Refinando tela {screen_id}...", Colors.CYAN)
        res = self._run_node_command("edit_screens", payload)
        if output_file and res.get("htmlCode"):
            self.save_screen_html(res, output_file)
        return res

    def get_screen(self, screen_id: str, output_file: Optional[str] = None) -> Dict[str, Any]:
        """Obtém detalhes, DOM HTML e screenshot de uma tela."""
        project_id = self.project_id or require_env("STITCH_PROJECT_ID")
        payload = {
            "projectId": project_id,
            "screenId": screen_id,
            "name": f"projects/{project_id}/screens/{screen_id}"
        }
        log("STITCH", f"Consultando detalhes da tela {screen_id}...", Colors.CYAN)
        res = self._run_node_command("get_screen", payload)
        if output_file and res.get("htmlCode"):
            self.save_screen_html(res, output_file)
        return res

    def list_screens(self, project_id: Optional[str] = None) -> List[Dict[str, Any]]:
        """Lista todas as telas criadas no projeto Stitch ativo."""
        proj_id = project_id or self.project_id or require_env("STITCH_PROJECT_ID")
        payload = {"projectId": proj_id}
        log("STITCH", f"Listando telas do projeto {proj_id}...", Colors.CYAN)
        res = self._run_node_command("list_screens", payload)
        if isinstance(res, dict) and "screens" in res:
            return res["screens"]
        if isinstance(res, list):
            return res
        return []

    def get_project(self, project_id: Optional[str] = None) -> Dict[str, Any]:
        """Obtém metadados do projeto e do Design System no Stitch."""
        return get_project_core(self, project_id)

    def create_project(self, title: Optional[str] = None) -> Dict[str, Any]:
        """Cria um novo projeto/workspace no Stitch SDK."""
        return create_project_core(self, title)

    def list_projects(self, filter_view: Optional[str] = None) -> List[Dict[str, Any]]:
        """Lista todos os projetos Stitch acessíveis ao usuário."""
        return list_projects_core(self, filter_view)


    def generate_variants(
        self,
        screen_id: str,
        prompt: str,
        variant_count: int = 3,
        creative_range: str = "EXPLORE",
        aspects: Optional[List[str]] = None,
        device_type: Optional[str] = None,
        model_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """Gera variantes visuais exploratórias a partir de uma tela."""
        project_id = self.project_id or require_env("STITCH_PROJECT_ID")
        resolved_device = (device_type or get_device_type())
        payload = {
            "projectId": project_id,
            "selectedScreenIds": [screen_id],
            "prompt": prompt,
            "variantOptions": {
                "variantCount": variant_count,
                "creativeRange": creative_range
            }
        }
        if aspects:
            payload["variantOptions"]["aspects"] = aspects
        if resolved_device:
            payload["deviceType"] = resolved_device
        if model_id:
            payload["modelId"] = model_id

        log("STITCH", f"Gerando {variant_count} variantes para a tela {screen_id}...", Colors.CYAN)
        return self._run_node_command("generate_variants", payload)

    def download_assets(self, output_dir: str, project_id: Optional[str] = None) -> Dict[str, Any]:
        """Baixa todas as telas e assets do projeto Stitch para um diretório local."""
        return download_assets_core(self, output_dir, project_id)

    def upload_asset(self, file_path: str, project_id: Optional[str] = None) -> Dict[str, Any]:
        """Faz upload de asset visual ou documento para o projeto Stitch."""
        return upload_asset_core(self, file_path, project_id)

    def create_design_system(self, design_system: Dict[str, Any], project_id: Optional[str] = None) -> Dict[str, Any]:
        """Cria um novo Design System no Stitch."""
        return create_design_system_core(self, design_system, project_id)

    def update_design_system(self, asset_name: str, design_system: Dict[str, Any], project_id: Optional[str] = None) -> Dict[str, Any]:
        """Atualiza um Design System existente no Stitch."""
        return update_design_system_core(self, asset_name, design_system, project_id)

    def list_design_systems(self, project_id: Optional[str] = None) -> List[Dict[str, Any]]:
        """Lista todos os Design Systems associados ao projeto ativo."""
        return list_design_systems_core(self, project_id)

    def apply_design_system(
        self,
        asset_id: str,
        selected_screen_instances: List[Dict[str, str]],
        project_id: Optional[str] = None
    ) -> Dict[str, Any]:
        """Aplica o Design System às instâncias de tela selecionadas."""
        return apply_design_system_core(self, asset_id, selected_screen_instances, project_id)

    def sync_design_system(
        self,
        design_md_path: Optional[str] = None,
        design_system: Optional[Dict[str, Any]] = None,
        **overrides
    ) -> Dict[str, Any]:
        """Sincroniza arquivo design.md e tokens do projeto com o Design System oficial do Stitch."""
        return sync_design_system_core(self, design_md_path, design_system, **overrides)

    def call_tool(self, tool_name: str, payload: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Executa uma chamada direta e genérica a qualquer ferramenta do Stitch SDK."""
        p = dict(payload or {})
        if "projectId" not in p and self.project_id:
            p["projectId"] = self.project_id
        log("STITCH", f"Invocando tool '{tool_name}' no Stitch SDK...", Colors.CYAN)
        return self._run_node_command(tool_name, p)


# Funções utilitárias avulsas para import direto e compatibilidade retroativa
from amb_cli.integrations.stitch.stitch_core.helpers import (
    generate_screen,
    edit_screen,
    get_screen,
    list_screens,
    get_project,
    create_project,
    list_projects,
    generate_variants,
    download_assets,
    upload_asset,
    create_design_system,
    update_design_system,
    list_design_systems,
    apply_design_system,
    sync_design_system,
    call_tool,
)



